### ISSUE TITLE
Table of Content and Page Headers

### ISSUE THREAD

**Issue by kickingbull** [trade-informatics/caracal#24]
Is there a way to generate a Table of Contents as well as a page header that is repeated on each page?

**Comment by jdugan** [comment]
Neither of these features is currently supported.  The main README is fairy exhaustive.  If you don't see something there or demonstrated in the example project, odds are Caracal doesn't do it right now.

A TOC would obviously be a great feature.  This is, however, a non-trivial effort.  The project that necessitated Caracal did not require a TOC (it was really more of a report builder thing), so a TOC did not make it into the MVP.  

A persistent header would likely be somewhat easier to implement.  I presume it would require some sort of document-level directive to accumulate the required content, coupled with a new renderer to construct a header.xml file.

As I hope is evident from the other issues on the project, I try to be responsive to all concerns (and will even tack on a new feature on request when doing so is super simple), but these two requests are a bit outside the scope of what I can manage personally right now.  If you'd like to fork the project and tackle either feature, that would be wonderful.  I'd be happy to answer any questions you have along the way.

**Comment by kickingbull** [comment]
Hi John,

Thanks for your response.  I kind of figured the TOC wasn’t supported yet and agree it is not a trivial thing to implement.

I may take a shot at the persistent header once I finish with some current work and time permitting.  I’ll followup with you if I get a chance to start looking into it.

Thanks again and great job on what you’ve implemented so far.

Keith

**Comment by cfacello** [comment]
Hi, I have a ~~dirty/temporal~~ solution for the missing TOC feature!  Here is what I did:

1. Create a new docx with only a TOC in the first page (obviously, it will be empty)
2. Import it into your generated doc using the _iframe_ feature like this:
`docx.iframe data: File.read('app/assets/docx_generation/index.docx')`
3. Profit!

For better, the TOC in the generated document, will be updated!

**Comment by ahmedelmahalawey** [comment]
@cfacello I have tried this solution and it creates the TOC but it doesn't auto update when I open the generated file, I had to right click on the TOC and press update fields. So is there some magic I should do so it can update automatically?

**Comment by cfacello** [comment]
@ahmedelmahalawey interesting... Just to know, have you tried opening it with Microsoft Word or with Open Office editor?  I only tried in the second one and it worked.

**Comment by ahmedelmahalawey** [comment]
Yeah I think this is the issue I have opened it with Microsoft Word and unfortunately I need it to be working with Microsoft Office. Thanks anyways

**Comment by jdugan** [comment]
I know you didn't submit a PR here, but as a point of reference for anyone reading this far, all Caracal features MUST work in Word to be accepted.  It's always a pleasant bonus if the feature CAN work in OpenOffice, LibreOffice, etc., but because Word is a Microsoft format, that's the baseline.

**Comment by goulvench** [comment]
I built this functionality in [my fork](https://github.com/goulvench/caracal/tree/add-table-of-contents), will be happy to build a PR to add this to core. Not sure it works as smoothly as should be, but I did my best. Please let me know how this should be called from templates, and what options should be provided if you're not satisfied with [the way I implemented this](https://github.com/goulvench/caracal/tree/add-table-of-contents#table-of-contents).

**Comment by jamesridgway** [comment]
If anyone is interested/still needs this, we have a fork that we're maintaining that has a working implementation of a table of contents:
https://github.com/the-curve-consulting/caracal

It also resolves the rubyzip issue and has a few other bits such as header/footer page numbers etc
